ERCS/erc-7774.md

```diff
@@ -0,0 +1,178 @@
+
+#### Handling Requests in "Listening for Events" State
+
+When a request is received in the "Listening for Events" state:
+
+1. **If no mapping entry exists**:
+   - The `web3://` client queries the smart contract.
+   - If the response includes the `evm-events` directive in the `Cache-Control` header and an `ETag`, a mapping entry is created using the `ETag`.
+   - If the response contains the `evm-events` directive and a `max-age=XX` directive in the `Cache-Control` header, the mapping entry is created with the `last modified date`, determined in the following order of priority:
+     - The `Last-Modified` header, if present.
+     - The `Date` header, if present.
+     - Otherwise, the block date when the smart contract was queried.
+   - If the response includes both an `ETag` and a `Cache-Control: evm-events max-age=XX` directive, a single mapping entry is created containing both the `ETag` and the `last modified date`.
+
+2. **If a mapping entry exists**:
+   - If the request contains a valid `If-None-Match` header:
+     - If the `ETag` in the mapping matches the `If-None-Match` value, the `web3://` client returns a `304 Not Modified` response immediately.
+     - If the `ETag` does not match, the client queries the smart contract, deletes the mapping entry, and processes the request as if no mapping entry existed.
+   
+   - If the request contains a valid `If-Modified-Since` header:
+     - If the `last modified date` in the mapping is earlier than the `If-Modified-Since` date, the client returns a `304 Not Modified` response immediately.
+     - Otherwise, the client queries the smart contract, deletes the mapping entry, and processes the request as if no mapping entry existed.
+   
+   - If the request contains neither `If-None-Match` nor `If-Modified-Since` headers (or they are invalid):
+     - The client queries the smart contract, deletes the mapping entry, and processes the request as if no mapping entry existed.
+
+
+
+#### Cache Invalidation via Blockchain Events
+
+In the "Listening for Events" state, the `web3://` client listens to the blockchain for the cache invalidation events defined in the previous section. For each path match, it deletes the corresponding mapping entry.
+
+
+## Rationale
+
+We add this feature to the [ERC-6944](./eip-6944.md) resolve mode because it can be added without changes the interface.
+
+To stay as close as possible to standard HTTP, we reuse the HTTP caching mechanism headers.
+
+The use of the `evm-events` directive is necessary to avoid a situation where a website uses traditional [RFC 9111](https://www.rfc-editor.org/rfc/rfc9111) HTTP caching headers, but does not implement this ERC  by failing to emit the events. In such cases, `web3://` clients implementing this ERC would serve stale content for that website indefinitely.
+
+## Security Considerations
+
+No security considerations were found.
```

Comment:
There are definitely some security considerations, no? At the very least contracts need to correctly emit cache invalidation events correctly, or stale content could be served.

I bet there are some unpleasant interactions between this and chain reorgs. as well?

Reply:
Yes indeed! 
I have added 3 : 

- Stale content will be served during the delay between a user transaction emitting a cache clearing event, and the `web3://` client picking and processing the event.
- For each cached page, websites must properly implement cache invalidation events; otherwise, stale content will be served indefinitely.
- In the event of a chain reorganization, the `web3://` client must roll back its caching state, or reverted content will be served until the next cache clearing event.

The first one is something that conceptually annoys me (even if the `web3://` client is optimizing its logs reading by using streams, it can theorically be missed).
So I am thinking about ways for the website to signal to the `web3://` client when fetching a page: "hey please ensure you are up to date with block XXX (because I just made a change on it via a TX)".
A HTTP header in the request is an idea, but may be not practical when doing a page change (e.g. in blog post edit page, save tx, then go to blog post page view). Thinking...